openlibrary/plugins/openlibrary/home.py

```diff
@@ -47,48 +49,106 @@ def GET(self):
             returncart_list=returncart_list,
             user=user, loans=loans)
 
-@public
-def popular_carousel(limit=36):
-    """popular works across lists which are available for reading
-    (borrowable or downloadable; not daisy only)
 
-    Limit is 36 by default to show 6 pages of the 6-item carousels on
-    landing page.
+def get_popular_books(seeds, limit=None, loan_check_batch_size=50,
+                      available_only=True):
+    """Takes a list of seeds (from user lists or otherwise), checks which
+    ones are (un)available (in batch sizes of
+    `loan_check_batch_size`), fetches data for editions (according to
+    available_only), and returns the tuple of available and
+    unavailable popular works.
+
+    Args:
+        seeds (list) - seeds from a user's list
+
+        limit (int) - Load the popular carousel with how many items?
+        (preferably divisible by 6; number of books shown per page)
 
-    lst1 is a manually curated collection of popular available books
-    which was constructed by looking at goodreads
-    (http://www.goodreads.com/list/show/1.Best_Books_Ever) Best Ever
-    list.
+        loan_check_batch_size (int) - Bulk submits this many
+        archive.org itemids at a time to see if they are available to
+        be borrowed (only considers waitinglist and bookreader
+        borrows, no acs4)
 
-    lst2 comes from the "top 2000+ most requested print disabled
-    eBooks in California" displayed from the /lists page.
+        available (bool) - return only available works? If True,
+        waitlisted_books will be []
 
-    Because lst1 is more highly curated and has more overall
-    recognizable and popular books, we prioritize drawing from this
-    list (randomized) and then fallback to lst2 (as lst1 is depleted
-    by virtue of loans).
+    Returns:
+        returns a tuple (available_books,
+        waitlisted_books). waitlisted_books will be [] if
+        available_only=True
 
     """
-    books = []
-    lst1 = web.ctx.site.get('/people/mekBot/lists/OL104041L')
-    lst2 = web.ctx.site.get('/people/openlibrary/lists/OL104411L')
-    if lst1 and lst2:
-        seeds1, seeds2 = lst1.seeds, lst2.seeds
-        random.shuffle(seeds1)
-        random.shuffle(seeds2)
-        seeds = seeds1 + seeds2
-        while seeds and len(books) < limit:
+    available_books = []
+    waitlisted_books = []
+
+    batch = {}
+    while seeds and (limit is None or len(available_books) < limit):
+        archive_ids = []
+        while seeds and len(batch) < loan_check_batch_size:
             seed = seeds.pop(0)
             key = seed['key']
-            ebook = seed.get_ebook_info()
-            if 'daisy_url' in ebook and 'borrow_url' not in ebook:
```

Comment:
We may still want to do this check to guarantee no daisy makes it on the list (which may be true for California list books)